crud/common/sharding_key.lua

```diff
@@ -0,0 +1,147 @@
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, WrongShardingConfigurationError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- Return a map or nil when metadata is not available.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return nil
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    local metadata_map = {}
+    for _, tuple in sharding_key_space:pairs() do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local sharding_key_def = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        local space_format = box.space[space_name]:format()
+        metadata_map[space_name] = {
+            sharding_key_def = sharding_key_def,
+            space_format = space_format,
+        }
+    end
+
+    return metadata_map
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch metadata from
+-- storages. It is not good from performance point of view. reentrant_fetch()
+-- wraps a fetch_on_storage() to limit a number of fibers that fetches a
+-- metadata by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch metadata
+-- if a cache is still empty.
+local function reentrant_fetch()
+    if cache.sharding_key_as_index_obj_map ~= nil then
+        return cache.sharding_key_as_index_obj_map
+    end
+
+    local timeout_deadline = fiber.clock() + const.FETCH_SHARDING_KEY_TIMEOUT
+    local timeout_is_exceeded
+    while sharding_key_fetch_in_progress == true do
+        timeout_is_exceeded = false
+        local timeout = timeout_deadline - fiber.clock()
+        if sharding_key_fetch_cond:wait(timeout) then
+            timeout_is_exceeded = true
+        end
+        if cache.sharding_key_as_index_obj_map ~= nil then
+            return cache.sharding_key_as_index_obj_map
+        end
+        if timeout_is_exceeded == true then
+            return nil, FetchShardingKeyError:new(
+                            'Waiting for sharding key to be fetched is timed out (%d sec)',
+                            const.FETCH_SHARDING_KEY_TIMEOUT)
+        end
+    end
+
+    sharding_key_fetch_in_progress = true -- Start of critical section.
+    local metadata_map, err = call.any(FETCH_FUNC_NAME, {}, {
+        timeout = timeout_deadline - fiber.clock()
+    })
+    if err ~= nil then
+        sharding_key_fetch_in_progress = false -- End of critical section.
+        sharding_key_fetch_cond:broadcast()
+        return nil, err
+    end
+
+    -- metadata_map is nil when fetch_on_storage() is timed out.
```

Comment:
Will not net.box raise an error in case of timeout?

```lua
tarantool> box.cfg{listen = 3301}
-- grants are already given
tarantool> net_box = require('net.box')
tarantool> c = net_box.connect(3301)
tarantool> fiber = require('fiber')
tarantool> pcall(c.call, c, 'fiber.sleep', {10}, {timeout = 1})
---
- false
- Timeout exceeded
...
```

Or it is wrapped by vshard?

Reply:
Even if it is wrapped, I guess, `err` will not be `nil` in the case. So the comment is strange. The code looks as handling of lack of the _ddl_sharding_key space (my guess).

Reply:
`fetch_on_storage()` returns `nil` when `box.space._ddl_sharding_key` is absent and table otherwise.
`fetch_on_storage()` wrapped by `call.any` (`crud/common/call.lua`) that wraps vshard's `call()` (see [API documentation](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_api/#lua-function.replicaset_object.call)). `replicaset_object:call` return result of `function_name` on success and `nil`, err otherwise.
Current version in `_fetch_on_router()` handle result as well as err from `call.any`:
```lua
    local metadata_map, err = call.any(FETCH_FUNC_NAME, {}, {                    
        timeout = timeout                                                        
    })                                                                           
    if err ~= nil then                                                           
        return err                                                               
    end                                                                          
    if metadata_map == nil then                                                  
        cache.sharding_key_as_index_obj_map = {}                                 
        return                                                                   
    end 
```

Comment `    -- metadata_map is nil when fetch_on_storage() is timed out.` has been removed, because it duplicates the description of `fetch_on_storage()`.